inference-engine/src/transformations/src/transformations/op_conversions/einsum_decomposition.cpp

```diff
@@ -0,0 +1,636 @@
+    // to the common, separate and reduced dimensions are placed in the correct order
+    if (is_separate_first) {
+        new_shape_parts.insert(new_shape_parts.end(), separate_parts.begin(), separate_parts.end());
+        new_shape_parts.insert(new_shape_parts.end(), reduced_sub_shape_prod.begin(), reduced_sub_shape_prod.end());
+    } else {
+        new_shape_parts.insert(new_shape_parts.end(), reduced_sub_shape_prod.begin(), reduced_sub_shape_prod.end());
+        new_shape_parts.insert(new_shape_parts.end(), separate_parts.begin(), separate_parts.end());
+    }
+
+    // in case of scalar reshape is not needed
+    if (new_shape_parts.size() == 0) {
+        return input_node;
+    }
+
+    // concatenate parts of new shape
+    auto new_shape_op = std::make_shared<ngraph::opset7::Concat>(new_shape_parts, 0);
+
+    auto reshaped_input_op = std::make_shared<ngraph::opset7::Reshape>(input_node, new_shape_op->output(0), false);
+
+    subgraph_nodes.insert(subgraph_nodes.end(), {new_shape_op, reshaped_input_op});
+    return reshaped_input_op->output(0);
+}
+
+/// \brief      Transpose one of the Einsum inputs to layout specified through the required
+/// subscript
+///
+/// \param      input_nodes         A vector of input nodes to Einsum
+/// \param      input_subscripts    A vector of corresponding subscripts for input nodes
+/// \param      required_subscript  The required subscript that defines layout to which the
+/// input is to transpose \param      input_ind           An index of the input node to be
+/// transposed \param      subgraph_nodes      A vector of operation nodes that is included into
+/// a sub-graph decomposing Einsum that is needed for copy_runtime_info
+///
+void transpose_input(ngraph::OutputVector& input_nodes, std::vector<std::string>& input_subscripts, const std::string& required_subscript, size_t input_ind,
+                     ngraph::NodeVector& subgraph_nodes) {
+    // perform sanity check for arguments
+    auto num_inputs = input_nodes.size();
+    NGRAPH_CHECK(num_inputs == input_subscripts.size(), "Each input must have own subscript.");
+    NGRAPH_CHECK(input_ind < num_inputs, "Input index is out of range.");
+
+    // generate permutation vector by searching for bijection between input_subscripts and
+    // required_subscript
+    std::vector<int64_t> permutation;
+    const auto& input_subscript = input_subscripts[input_ind];
+
+    // transpose is not needed since the input subscript is not going to be changed
+    if (required_subscript == input_subscript) {
+        return;
+    }
+
+    // find permutation that establishes bijection between the input subscript and the required
+    // one
+    auto labels = ngraph::op::v7::Einsum::extract_labels(input_subscript);
+    auto required_labels = ngraph::op::v7::Einsum::extract_labels(required_subscript);
+    NGRAPH_CHECK(labels.size() == required_labels.size());
+    for (const auto& required_label : required_labels) {
```

Comment:
Yes, you are correct. Hence, we can re-use it instead of compute it again. But costs for this shouldn't be significant since its length is just a tensor rank. Also this routine is used by parts for which `generate_subscript` is not called. So I decided to unify this transpose for all cases by use of subscript.